internal/fxlog/fxlog.go

```diff
@@ -21,68 +21,121 @@
 package fxlog
 
 import (
-	"fmt"
-	"log"
 	"os"
-	"strings"
 
 	"go.uber.org/fx/internal/fxreflect"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
 var _exit = func() { os.Exit(1) }
 
-// Printer is a formatting printer.
-type Printer interface {
-	Printf(string, ...interface{})
+// Level is the level of logging used by Logger.
+type Level int
+
+const (
+	InfoLevel = iota
+	ErrorLevel
+)
+
+// Entry is an entry
+type Entry struct {
+	Level   Level
+	Message string
+	Fields  []Field
+	Stack   string
 }
 
-// New returns a new Logger backed by the standard library's log package.
-func New() *Logger {
-	return &Logger{log.New(os.Stderr, "", log.LstdFlags)}
+func (e Entry) WithStack(stack string) Entry {
+	e.Stack = stack
+
+	return e
 }
 
-// A Logger writes output to standard error.
-type Logger struct {
-	Printer
+func (e Entry) Write(logger Logger) {
+	logger.Log(e)
 }
 
-// Printf logs a formatted Fx line.
-func (l *Logger) Printf(format string, v ...interface{}) {
-	l.Printer.Printf(prepend(format), v...)
+type Field struct {
+	Key   string
+	Value interface{}
 }
 
-// PrintProvide logs a type provided into the dig.Container.
-func (l *Logger) PrintProvide(t interface{}) {
-	for _, rtype := range fxreflect.ReturnTypes(t) {
-		l.Printf("PROVIDE\t%s <= %s", rtype, fxreflect.FuncName(t))
+type Logger interface {
+	Log(entry Entry)
+	PrintProvide(interface{})
+	PrintSupply(interface{})
```

Comment:
We should try to keep the interface small. PrintProvide and PrintSupply can be implemented in terms of Logger.Log. In fact, the implementation will likely always be the same. So we can make them top-level functions that operate on a logger rather than methods of the logger:

```
func PrintProvide(Logger, interface{})
func PrintSupply(Logger, interface{})
```